exchangelib/credentials.py

```diff
@@ -70,15 +103,50 @@ def __str__(self):
 
 @python_2_unicode_compatible
 class OAuth2Credentials(BaseCredentials, PickleMixIn):
-    """Login info for OAuth 2.0 authentication
+    """
+    Login info for OAuth 2.0 client credentials authentication, as well
+    as a base for other OAuth 2.0 grant types.
+
+    This is primarily useful for in-house applications accessing data
+    from a single Microsoft account. For applications that will access
+    multiple tenants' data, the client credentials flow does not give
+    the application enough information to restrict end users' access to
+    the appropriate account. Use OAuth2AuthorizationCodeCredentials and
+    the associated auth code grant type for multi-tenant applications.
+
+    :param client_id: ID of an authorized OAuth application
+    :param client_secret: Secret associated with the OAuth application
+    :param tenant_id: Microsoft tenant ID of the account to access
     """
     __slots__ = ('client_id', 'client_secret', 'tenant_id')
 
     def __init__(self, client_id, client_secret, tenant_id):
+        super(OAuth2Credentials, self).__init__()
         self.client_id = client_id
         self.client_secret = client_secret
         self.tenant_id = tenant_id
 
+    def refresh(self):
+        # Creating a new session gets a new access token, so there's no
+        # work here to refresh the credentials. This implementation just
+        # makes sure we don't raise a NotImplementedError.
+        pass
+
+    def on_token_auto_refreshed(self, access_token):
+        """
+        Called after the access token is refreshed (requests-oauthlib
+        can automatically refresh tokens if given an OAuth client ID and
+        secret, so this is how our copy of the token stays up-to-date).
+        Applications that cache access tokens can override this to store
+        the new token - just remember to call the super() method!
```

Comment:
This is slightly non-ideal - I can't think of a situation where applications would want to _override_ the default behavior so much as be _notified_ about the token update, and it's clunky to need to subclass just to add what's effectively an event handler. I would've just gone with `observable` if it were already a dependency, but since it's not and I didn't want to re-implement it, I went with this. What do you think - would `observable` be an acceptable new dependency, or is it not worth adding for this one case?